p2p_beef_tx.go

```diff
@@ -114,68 +70,138 @@ func DecodeBEEF(beefHex string) (*DecodedBEEF, error) {
-
 	transactions = transactions[:len(transactions)-1]
 
 	return &DecodedBEEF{
-		CMPSlice:        cmpSlice,
+		BUMPs:           bumps,
 		InputsTxData:    transactions,
 		ProcessedTxData: processedTx,
 	}, nil
 }
 
-func decodeCMPSliceFromStream(hexBytes []byte) (CMPSlice, []byte, error) {
-	if len(hexBytes) == 0 {
-		return nil, nil, errors.New("cannot decode cmp slice from stream - no bytes provided")
+func decodeBUMPs(beefBytes []byte) ([]BUMP, []byte, error) {
+	if len(beefBytes) == 0 {
+		return nil, nil, errors.New("cannot decode BUMP - no bytes provided")
 	}
 
-	nCMPs, bytesUsed := bt.NewVarIntFromBytes(hexBytes)
-	hexBytes = hexBytes[bytesUsed:]
+	bumps := make([]BUMP, 0)
+	nBump, bytesUsed := bt.NewVarIntFromBytes(beefBytes)
+	beefBytes = beefBytes[bytesUsed:]
 
-	var cmpPaths []CompoundMerklePath
-	for i := 0; i < int(nCMPs); i++ {
-		cmp, bytesUsedToDecodeCMP, err := NewCMPFromStream(hexBytes)
+	for i := 0; i < int(nBump); i++ {
+		if len(beefBytes) < 1 {
+			return nil, nil, errors.New("insufficient bytes to extract BUMP blockHeight")
+		}
+		blockHeight, bytesUsed := bt.NewVarIntFromBytes(beefBytes)
+		beefBytes = beefBytes[bytesUsed:]
+		bumpPaths, remainingBytes, err := decodeBUMPPathsFromStream(beefBytes)
 		if err != nil {
 			return nil, nil, err
 		}
+		beefBytes = remainingBytes
 
-		cmpPaths = append(cmpPaths, cmp)
-		hexBytes = hexBytes[bytesUsedToDecodeCMP:]
-	}
+		bump := BUMP{
+			blockHeight: uint64(blockHeight),
+			path:        bumpPaths,
+		}
 
-	cmpSlice := CMPSlice(cmpPaths)
+		bumps = append(bumps, bump)
+	}
 
-	return cmpSlice, hexBytes, nil
+	return bumps, beefBytes, nil
 }
 
-func NewCMPFromStream(hexBytes []byte) (CompoundMerklePath, int, error) {
-	height, bytesUsed, err := extractHeight(hexBytes)
-	if err != nil {
-		return nil, 0, err
+func decodeBUMPPathsFromStream(hexBytes []byte) ([][]BUMPLeaf, []byte, error) {
+	if len(hexBytes) < 1 {
+		return nil, nil, errors.New("cannot decode BUMP paths from stream - no bytes provided")
 	}
+
+	treeHeight, bytesUsed := bt.NewVarIntFromBytes(hexBytes)
 	hexBytes = hexBytes[bytesUsed:]
+	bumpPaths := make([][]BUMPLeaf, 0)
 
-	var cmp CompoundMerklePath
-	currentHeight := height
-	bytesUsedToDecodeCMP := bytesUsed
+	for i := 0; i < int(treeHeight); i++ {
+		if len(hexBytes) < 1 {
+			return nil, nil, errors.New("cannot decode BUMP paths number of leaves from stream - no bytes provided")
+		}
+		nLeaves, bytesUsed := bt.NewVarIntFromBytes(hexBytes)
+		hexBytes = hexBytes[bytesUsed:]
+		bumpPath, remainingBytes, err := decodeBUMPPath(nLeaves, hexBytes)
+		if err != nil {
+			return nil, nil, err
+		}
+		hexBytes = remainingBytes
+		bumpPaths = append(bumpPaths, bumpPath)
+	}
 
-	for currentHeight >= 0 {
-		var pathMap map[string]uint64
+	return bumpPaths, hexBytes, nil
+}
 
-		pathMap, bytesUsed, err = extractPathMap(hexBytes, currentHeight)
-		if err != nil {
-			return nil, 0, err
+func decodeBUMPPath(nLeaves bt.VarInt, hexBytes []byte) ([]BUMPLeaf, []byte, error) {
```

Comment:
**Minor**: maybe call it `decodeBUMPLevel`? WDYT?

Reply:
Renamed